### PR TITLE
Added enrollment question check to CourseModulePermissionBase

### LINKED ISSUE #679: Enrollment questionnaire is affected by the module opening time
Continued from https://github.com/apluslms/a-plus/issues/462

Enrollment questionnaires should not be affected by the module opening time. The course already has its own enrollment opening/closing time settings and they are enough. Enrollment questionnaires should not be affected by the module opening/closing times. Fixing this is difficult because the `CourseModulePermission` does not directly have access to the exercise and it can't check if the exercise is an enrollment exercise. Perhaps the exercise could be defined in the attributes of the `view` object and it would always be defined if the user is on an exercise page. If the exercise is defined and it is an enrollment exercise, then the course module opening time is ignored in the permission.

https://github.com/apluslms/a-plus/blob/cd10e04963557444efeb11b06b2834bf47ace903/course/permissions.py#L140

### PR DESCRIPTION
**Pull request by EerikSaksi**
Before, the permission check would work by simply checking whether the module had opened yet. With this check, the module opening time is only checked if the exercise is not an enrollment question

Fixes #679

# Description

**What?**

This change adds a check for if the exercise is an enrolment question to CourseModulePermissionBase.

**Why?**

Enrolment questions should be accessible before the module is open

**How?**

Before the check for if the module is open, we check whether the exercise is an enrolment question. If this is the case, then we do not care if the module is open.

Fixes #679 

# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [ ] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [X] Manual testing.

Ensuring that students can open the URL to enrolment exercises in modules which have not opened yet, making sure that no error is thrown when the view supplied to CourseModulePermissionBase does not have a exercise attribute (when called from ModuleView and not ExerciseView)

**Did you test the changes in**

- [ ] Chrome
- [X] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [ ] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [ ] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: course/permissions.py [hunk 1]
```diff
@@ -248,7 +248,8 @@ def has_object_permission(self, request: HttpRequest, view: Any, module: CourseM
         if module.status == CourseModule.STATUS.HIDDEN:
             return False
 
-        if not module.is_after_open():
+        #if this isn't an enrollment question and the module hasn't opened yet
+        if not (hasattr(view, 'exercise') and view.exercise.status in ('enrollment', 'enrollment_ext')) and not module.is_after_open():
```
Comment (PasiSa) [inline course/permissions.py:252]: Should we rather use
```
view.exercise.status in (LearningObject.STATUS.ENROLLMENT, LearningObject.STATUS.ENROLLMENT_EXTERNAL)
```
Perhaps could also be split into multiple lines, as this line is getting rather long

**Review by PasiSa — CHANGES_REQUESTED** [review]
One style-related proposal, otherwise looks good to me. Didn't test this PR (or my proposal), though.

**Comment by EerikSaksi** [comment]
Changed the formatting and added LearningObject import @PasiSa

**Review by PasiSa — APPROVED** [review]

**Review by Mankro — CHANGES_REQUESTED** [review]
Good start! When I tested this, I noticed that it really does not work yet because there is also another permission (`ExerciseVisiblePermissionBase`) that checks the module opening time. Fixing that other permission is easy. I also thought that some details could be formatted better. `CourseModulePermissionBase.has_object_permission()` has the part `module.requirements.exists()` in the end that should not be used in enrollment questionnaires.

I made some changes myself and pushed them to git so that you may finalize this PR.

**git commit message**
* commit message heading: use the imperative mood
* line length: at most 80 characters
* the heading could be reworded better. Emphasize the goal ("enrollment questionnaires should not be affected by the module opening time") instead of the implementation details on the code lines ("check in CourseModulePermissionBase").

**Comment by EerikSaksi** [comment]
Thanks @Mankro for the feedback, I will get to it!

**Comment by EerikSaksi** [comment]
I tested the new changes, it seems like the enrolment questionnaire is accessible before the course is open, but the course is not. I am not sure if this is the desired behaviour.

https://user-images.githubusercontent.com/46625399/191976229-87631e78-d401-4e5f-bb4a-ef38aaccd975.mp4

**Comment by Mankro** [comment]
> > it seems like the enrolment questionnaire is accessible before the course is open, but the course is not. I am not sure if this is the desired behaviour.
> 
> I haven’t studied this issue / PR properly at all, but I’ll just mention that AFAIC, it shouldn’t be possible to fill in or even see the enrollment questionnaire before the course is actually open.

If I remember correctly, the enrollment questionnaire is supposed to be open if the course settings (enrollment starting time and enrollment ending time) define so. If the teacher wants to set enrollment starting time before the course opening time, then it is possible.

If the enrollment opening times are not defined, they default to the course opening times.

If the enrollments are fetched from Sisu, we should check what happens. Is access restricted by the course opening times?

**Comment by EerikSaksi** [comment]
I added a test for accessing the enrollment exercise which tests combinations of starting times and enrollment starting times which passes. I might have misunderstood what the expected behavior is but if the test describes the expected behavior then this should work.

**Review by Mankro — APPROVED** [review]
Good job with the unit tests! I added some more tests, rebased and now I am merging this.

**Pull request merged**